File: bit_life_survival/core/requirements.py
from __future__ import annotations

from typing import Any

from .loader import ContentBundle
from .models import GameState, has_owned_item, owned_item_ids


def _check_leaf(expr: dict[str, Any], state: GameState, content: ContentBundle) -> tuple[bool, list[str]]:
    if "hasTag" in expr:
        required_tag = expr["hasTag"]
        owned_ids = owned_item_ids(state)
        has_tag = any(required_tag in content.item_by_id[item_id].tags for item_id in owned_ids if item_id in content.item_by_id)
        if has_tag:
            return True, []
        return False, [f"Requires owned item with tag '{required_tag}'."]

    if "hasItem" in expr:
        item_id = expr["hasItem"]
        has_item = has_owned_item(state, item_id)
        return (True, []) if has_item else (False, [f"Requires item '{item_id}' in inventory or equipped slots."])

    if "flag" in expr:
        flag = expr["flag"]
        has_flag = flag in state.flags
        return (True, []) if has_flag else (False, [f"Requires flag '{flag}'."])

    if "meterGte" in expr:
        payload = expr["meterGte"]
        meter = payload["meter"]
        required_value = float(payload["value"])
        current = getattr(state.meters, meter)
        if current >= required_value:
            return True, []
        return False, [f"Requires {meter} >= {required_value}."]

    if "injuryLte" in expr:
        required_injury = float(expr["injuryLte"])
        if state.injury <= required_injury:
            return True, []
        return False, [f"Requires injury <= {required_injury}."]

    if "distanceGte" in expr:
        required_distance = float(expr["distanceGte"])
        if state.distance >= required_distance:
            return True, []
        return False, [f"Requires distance >= {required_distance}."]

    return False, ["Unknown requirement."]


def evaluate_requirement(expr: dict[str, Any], state: GameState, content: ContentBundle) -> tuple[bool, list[str]]:
    if "all" in expr:
        all_reasons: list[str] = []
        for child in expr["all"]:
            ok, reasons = evaluate_requirement(child, state, content)
            if not ok:
                all_reasons.extend(reasons)
        return (len(all_reasons) == 0, all_reasons)

    if "any" in expr:
        all_failures: list[str] = []
        for child in expr["any"]:
            ok, reasons = evaluate_requirement(child, state, content)
            if ok:
                return True, []
            all_failures.extend(reasons)
        if all_failures:
            return False, [f"Requires any of: {' OR '.join(all_failures)}"]
        return False, ["Requires any listed condition."]

    if "not" in expr:
        ok, _ = evaluate_requirement(expr["not"], state, content)
        if ok:
            return False, ["Blocked by negated requirement."]
        return True, []

    return _check_leaf(expr, state, content)

File: core/src/sim/requirements.ts
import { LoadedContent } from "../content/loader";
import { RequirementExpr } from "../content/schemas";
import { RunnerState, ownedItemIds, ownsItem } from "./state";

export interface RequirementCheck {
  readonly ok: boolean;
  readonly reasons: string[];
}

function evaluateLeaf(
  requirement: RequirementExpr,
  state: RunnerState,
  content: LoadedContent
): RequirementCheck {
  if ("hasTag" in requirement) {
    const ownedIds = ownedItemIds(state);
    const hasTag = ownedIds.some((itemId) => content.itemById[itemId]?.tags.includes(requirement.hasTag));
    return hasTag
      ? { ok: true, reasons: [] }
      : { ok: false, reasons: [`Requires an owned item tagged '${requirement.hasTag}'.`] };
  }
  if ("hasItem" in requirement) {
    const hasItem = ownsItem(state, requirement.hasItem);
    return hasItem
      ? { ok: true, reasons: [] }
      : { ok: false, reasons: [`Requires item '${requirement.hasItem}' in inventory or equipped.`] };
  }
  if ("flag" in requirement) {
    const hasFlag = state.flags.has(requirement.flag);
    return hasFlag ? { ok: true, reasons: [] } : { ok: false, reasons: [`Requires flag '${requirement.flag}'.`] };
  }
  if ("meterGte" in requirement) {
    const currentValue = state.meters[requirement.meterGte.meter];
    return currentValue >= requirement.meterGte.value
      ? { ok: true, reasons: [] }
      : {
          ok: false,
          reasons: [`Requires ${requirement.meterGte.meter} >= ${requirement.meterGte.value}.`]
        };
  }
  if ("injuryLte" in requirement) {
    return state.injury <= requirement.injuryLte
      ? { ok: true, reasons: [] }
      : { ok: false, reasons: [`Requires injury <= ${requirement.injuryLte}.`] };
  }
  if ("distanceGte" in requirement) {
    return state.distance >= requirement.distanceGte
      ? { ok: true, reasons: [] }
      : { ok: false, reasons: [`Requires distance >= ${requirement.distanceGte}.`] };
  }
  return { ok: false, reasons: ["Unknown requirement."] };
}

export function evaluateRequirement(
  requirement: RequirementExpr,
  state: RunnerState,
  content: LoadedContent
): RequirementCheck {
  if ("all" in requirement) {
    const checks = requirement.all.map((expr) => evaluateRequirement(expr, state, content));
    const failed = checks.filter((check) => !check.ok);
    return failed.length === 0
      ? { ok: true, reasons: [] }
      : { ok: false, reasons: failed.flatMap((check) => check.reasons) };
  }

  if ("any" in requirement) {
    const checks = requirement.any.map((expr) => evaluateRequirement(expr, state, content));
    if (checks.some((check) => check.ok)) {
      return { ok: true, reasons: [] };
    }
    const leafReasons = checks.flatMap((check) => check.reasons);
    return {
      ok: false,
      reasons: [
        leafReasons.length > 0
          ? `Requires any of: ${leafReasons.join(" OR ")}`
          : "Requires any of the listed requirements."
      ]
    };
  }

  if ("not" in requirement) {
    const check = evaluateRequirement(requirement.not, state, content);
    return check.ok
      ? { ok: false, reasons: ["Disallowed condition is currently true."] }
      : { ok: true, reasons: [] };
  }

  return evaluateLeaf(requirement, state, content);
}
